File: apua34/2회차/2302.극장좌석.js
/**
 * @file 2302.극장좌석.js
 * @author @apua34
 * @brief 
 * @link https://www.acmicpc.net/problem/2302
 * @version 0.1
 * @date 2023-07-23
 * @strategy DP
 *
 * @copyright Copyright (c) 2023 @apua34
 *
 */

const fs = require("fs");
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
const input = fs.readFileSync(filePath)
    .toString()
    .trim()
    .split("\n")

const [N, M, ...arr] = input;


function solution(N, M, arr) {
    const n = Number(N);
    const m = Number(M);

    if(n === m) return 1;

    let result = [];

    const vip = Array(n + 1).fill(false); // visited
    arr.map((target) => vip[target] = true);
    
    const fibo_memo = [0, 1, 2];
    
    for (let i = 3; i <= N - M; i++) {
        fibo_memo[i] = fibo_memo[i - 1] + fibo_memo[i - 2];
    }

    let current = 0;
    let value = 1;

    for(let i = 1; i <= N; i++) {

        if(!vip[i]) {
            value = fibo_memo[++current];
        } 
        
        if(vip[i]) {
            result.push(value);
            current = 0;
            value = 1;
        }
    }

    result.push(value);

   return result.reduce((acc, value) => acc * value, 1)
}

const check = solution(N, M, arr);
console.log(check)

File: Mootata/34회차/2302.극장 좌석.py
n = int(input())
m = int(input())
vip = [int(input()) for _ in range(m)]

dp = [0 for _ in range(n + 2)]
dp[0] = 1
dp[1] = 1
dp[2] = 2

for i in range(3, n + 1):
    dp[i] = dp[i - 1] + dp[i - 2]

answer = 1

if m > 0:
    pre = 0
    for j in range(m):  # vip 사이 그룹의 경우의 수
        answer *= dp[vip[j] - 1 - pre]
        pre = vip[j]
    answer *= dp[n - pre]
else:
    answer = dp[n]
print(answer)

# 1. 1 = 1
# 2. 1 2, 2 1 = 2
# 3. 1 2 3, 1 3 2, 2 1 3 = 3
# 4. 1 2 3 4, 1 3 2 4, 2 1 3 4, 1 2 4 3, 2 1 3 4 = 5
# 5. 1 2 3 4 5, 1 3 2 4 5, 2 1 3 4 5, 1 2 4 3 5, 2 1 3 4 5, 1 2 3 5 4, 1 3 2 5 4, 2 1 3 5 4 = 8

# dp[n] = dp[n - 1] + dp[n - 2]
